### ISSUE TITLE
Grails 3.1.1: Adding new methods to controller fails on reload with exception

### ISSUE THREAD

**Issue by jamie-kemp** [grails/grails-core#9686]
Adding new methods to a controller under run-app fails with the exception below. Modifying a pre-existing method appears to work fine. Initially thought it looked like a regression of #599, but trying to run under a 7 JDK still results in the exception.

Restarting run-app results in things working as expected. I have also tested with and without the new subproject/exploded functionality.

```
ERROR org.springsource.loaded.agent.Watcher - FileWatcher caught serious error, see cause
org.grails.core.exceptions.GrailsRuntimeException: Error instantiated artefact class [class my.app.InsideController] of type [class org.grails.core.DefaultGrailsControllerClass]: InvocationTargetException
        at grails.core.ArtefactHandlerAdapter.newArtefactClass(ArtefactHandlerAdapter.java:170) ~[grails-core-3.1.1.jar:3.1.1]
        at grails.core.DefaultGrailsApplication.addArtefact(DefaultGrailsApplication.java:784) ~[grails-core-3.1.1.jar:3.1.1]
        at grails.core.DefaultGrailsApplication.addArtefact(DefaultGrailsApplication.java:480) ~[grails-core-3.1.1.jar:3.1.1]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.7.0_95]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57) ~[na:1.7.0_95]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.7.0_95]
        at java.lang.reflect.Method.invoke(Method.java:606) ~[na:1.7.0_95]
        at org.springsource.loaded.ri.ReflectiveInterceptor.jlrMethodInvoke(ReflectiveInterceptor.java:1432) ~[springloaded-1.2.5.RELEASE.jar:1.2.5.RELEASE]
        at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoCachedMethodSite.invoke(PogoMetaMethodSite.java:169) ~[groovy-2.4.5.jar:2.4.5]
        at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.call(PogoMetaMethodSite.java:71) ~[groovy-2.4.5.jar:2.4.5]
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:133) ~[groovy-2.4.5.jar:2.4.5]
        at org.grails.plugins.web.controllers.ControllersGrailsPlugin.onChange(ControllersGrailsPlugin.groovy:189) ~[grails-plugin-controllers-3.1.1.jar:3.1.1]
        at org.grails.plugins.DefaultGrailsPlugin.notifyOfEvent(DefaultGrailsPlugin.java:792) ~[grails-core-3.1.1.jar:3.1.1]
        at org.grails.plugins.AbstractGrailsPluginManager.informOfClassChange(AbstractGrailsPluginManager.java:386) ~[grails-core-3.1.1.jar:3.1.1]
        at org.grails.dev.support.GrailsSpringLoadedPlugin.reloadEvent(GrailsSpringLoadedPlugin.java:52) ~[grails-core-3.1.1.jar:3.1.1]
        at org.springsource.loaded.TypeRegistry.fireReloadEvent(TypeRegistry.java:2022) ~[springloaded-1.2.5.RELEASE.jar:1.2.5.RELEASE]
        at org.springsource.loaded.ReloadableType.loadNewVersion(ReloadableType.java:444) ~[springloaded-1.2.5.RELEASE.jar:1.2.5.RELEASE]
        at org.springsource.loaded.TypeRegistry.loadNewVersion(TypeRegistry.java:1014) ~[springloaded-1.2.5.RELEASE.jar:1.2.5.RELEASE]
        at org.springsource.loaded.agent.ReloadableFileChangeListener.fileChanged(ReloadableFileChangeListener.java:104) ~[springloaded-1.2.5.RELEASE.jar:1.2.5.RELEASE]
        at org.springsource.loaded.agent.Watcher.determineChangesSince(FileSystemWatcher.java:251) [springloaded-1.2.5.RELEASE.jar:1.2.5.RELEASE]
        at org.springsource.loaded.agent.Watcher.run(FileSystemWatcher.java:235) [springloaded-1.2.5.RELEASE.jar:1.2.5.RELEASE]
        at java.lang.Thread.run(Thread.java:745) [na:1.7.0_95]
Caused by: java.lang.reflect.InvocationTargetException: null
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[na:1.7.0_95]
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57) ~[na:1.7.0_95]
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[na:1.7.0_95]
        at java.lang.reflect.Constructor.newInstance(Constructor.java:526) ~[na:1.7.0_95]
        at org.springsource.loaded.ri.ReflectiveInterceptor.jlrConstructorNewInstance(ReflectiveInterceptor.java:1075) ~[springloaded-1.2.5.RELEASE.jar:1.2.5.RELEASE]
        at grails.core.ArtefactHandlerAdapter.newArtefactClass(ArtefactHandlerAdapter.java:161) ~[grails-core-3.1.1.jar:3.1.1]
        ... 21 common frames omitted
Caused by: org.grails.core.exceptions.GrailsConfigurationException: Cannot find invokable controller action: test5
        at org.grails.core.DefaultGrailsControllerClass.methodStrategy(DefaultGrailsControllerClass.java:122) ~[grails-core-3.1.1.jar:3.1.1]
        at org.grails.core.DefaultGrailsControllerClass.<init>(DefaultGrailsControllerClass.java:67) ~[grails-core-3.1.1.jar:3.1.1]
        ... 27 common frames omitted
Caused by: java.lang.NoSuchMethodException: no such method: my.app.InsideController.test5()Object/invokeVirtual
        at java.lang.invoke.MemberName.makeAccessException(MemberName.java:758) ~[na:1.7.0_95]
        at java.lang.invoke.MemberName$Factory.resolveOrFail(MemberName.java:875) ~[na:1.7.0_95]
        at java.lang.invoke.MethodHandles$Lookup.resolveOrFail(MethodHandles.java:1133) ~[na:1.7.0_95]
        at java.lang.invoke.MethodHandles$Lookup.findVirtual(MethodHandles.java:696) ~[na:1.7.0_95]
        at org.grails.core.DefaultGrailsControllerClass.methodStrategy(DefaultGrailsControllerClass.java:119) ~[grails-core-3.1.1.jar:3.1.1]
        ... 28 common frames omitted
Caused by: java.lang.NoSuchMethodError: my.app.InsideController.test5()Ljava/lang/Object;
        at java.lang.invoke.MethodHandleNatives.resolve(Native Method) ~[na:1.7.0_95]
        at java.lang.invoke.MemberName$Factory.resolve(MemberName.java:847) ~[na:1.7.0_95]
        at java.lang.invoke.MemberName$Factory.resolveOrFail(MemberName.java:872) ~[na:1.7.0_95]
        ... 31 common frames omitted
```

**Comment by darcusfenix** [comment]
I also have this problem with

Grails Version: 3.1.1
Groovy Version: 2.4.5
JVM Version: 1.8.0_66

Linux Mint 17.2 Rafaela

**Comment by olavgg** [comment]
@graemerocher did you add this fix for 3.0.15? I still have this problem when I add new actions to a controller in the development environment.

**Comment by barcelona23** [comment]
I still have same problem but without exception. this only happens in OSx, using same project in windows works ok.

Grails Version: 3.1.11
| Groovy Version: 2.4.7
| JVM Version: 1.8.0_102

Any idea?
